Do not clear validationinterface entries being executed

sipa opened the pull request:
The previous code for MainSignalsInstance::Clear would decrement the reference
count of every interface, including ones that were already Unregister()ed but
still being executed.

This fixes the issue pointed out here: https://github.com/bitcoin/bitcoin/pull/18524/files#r404395685 . It's not currently observable.

sipa committed:
Do not clear validationinterface entries being executed

The previous code for MainSignalsInstance::Clear would decrement the reference
count of every interface, including ones that were already Unregister()ed but
still being executed.

fanquake commented:
I take it this resolves [this comment](https://github.com/bitcoin/bitcoin/pull/18524/files#r404395685)?

sipa commented:
@fanquake You're too fast. I was just looking to add a link where I made that comment.

promag commented:
This change is now covered in #18471.

ryanofsky committed:
Add test for UnregisterAllValidationInterfaces bug

Bug in MainSignalsInstance::Clear could cause validation interface callbacks to
be deleted during execution if UnregisterAllValidationInterfaces was called
more than once.

Bug was introduced in https://github.com/bitcoin/bitcoin/pull/18524 and is
fixed by https://github.com/bitcoin/bitcoin/pull/18551

ryanofsky commented:
Code review ACK c182fd517db2337f61b0038ca929fcfd01f18f07

I wrote a test for the bug that fails without the fix and passes with it: 4ba1627fb99ea46836e0d73ad1682b4b184124be ([branch](https://github.com/ryanofsky/bitcoin/commits/review.18551.1-edit))

Feel free to use the test here and squash into the fix commit.

ryanofsky commented:
Note: This PR fixes a bug in an internal function. The fix doesn't affect behavior of bitcoind, the gui, or any utilities because the function isn't called in a way that would trigger the bug

promag commented:
Code review ACK c182fd517db2337f61b0038ca929fcfd01f18f07.

sipa commented:
@ryanofsky Cherry-picked. Also updated the PR description to say it's not currently observable.

ryanofsky reviewed (approved):
Code review ACK 2276339a176f83ffe8ceefb3e41ecca8601aa13b. No change to bugfix, just rebased and new test commit added since last review

jonasschnelli commented:
utACK 2276339a176f83ffe8ceefb3e41ecca8601aa13b - reviewed code and test (thanks @ryanofsky for adding the test).

MarcoFalke commented:
ACK 2276339a176f83ffe8ceefb3e41ecca8601aa13b 🎎

<details><summary>Show signature and timestamp</summary>

Signature:

```
-----BEGIN PGP SIGNED MESSAGE-----
Hash: SHA512

ACK 2276339a176f83ffe8ceefb3e41ecca8601aa13b 🎎
-----BEGIN PGP SIGNATURE-----

iQGzBAEBCgAdFiEE+rVPoUahrI9sLGYTzit1aX5ppUgFAlwqrYAACgkQzit1aX5p
pUh20Qv+N8PfCJlrF51u9zKcnkK1GBJis/ghaoibar1FJ1i1HSbPRRL+eKYVv0Sy
1EtNyAVItG33UTdTRWO5ektfVHDweRO3XZZwknk0Jo/iUo+dLMhXoEvu4/izJ09V
3K/ae/hZDnX5rARkOIW00NYtFdm3ynWtPXG6g5p5U9RJeZ31StmJoXp8FESDRhsl
qx5vJSEhV7hG+oUl6seSJLFdMZ3F8E6/OPg7kBMCRPKJI9/CrKrbUMLUbiV8LZy3
RJFrcCBWWa0Nw5T+mxoHl83iHJZKnPK5cKLzNYt1IkBC826KJKE1sN5lPz6zReug
cMkr3ycS8TlXiNRo8fsQNaCV7xj0qAVcXgqqJurOyA0DadMSaddvj9NMD93BZDI+
xO6fY8bVY6c4yTY+le0Ap0LhTdtvK4w//ld8K2lPMEVuhH3m5e8vjJHW87VWQQiG
xL78hivW6rda/QqCKs4Ad5CupVLrCjVpDcCu2dZihoXmuqpXmH9Dekz2vdbR4/Il
nKUooeqh
=fz/C
-----END PGP SIGNATURE-----
```

[Timestamp of file with hash `85ad0a848e9b8fccb5f691da9f517f86522c78158bd9a4bd80a6403b8f07cf8b  -`](https://opentimestamps.org/info/?004f70656e54696d657374616d7073000050726f6f6600bf89e2e884e89294010885ad0a848e9b8fccb5f691da9f517f86522c78158bd9a4bd80a6403b8f07cf8bf010c1f0df3b92bef0cf0d2c77cbab15649808fff0100a41d85e89ad84d87a103b4c2e77617008f1045e8dc688f00828053e7c372f19a90083dfe30d2ef90c8e2e2d68747470733a2f2f616c6963652e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267fff0108957458cd650d325dad212526081b27908f1045e8dc689f008a2b27b1761addc9e0083dfe30d2ef90c8e232268747470733a2f2f6274632e63616c656e6461722e636174616c6c6178792e636f6dfff010d080e0a69bbba084108627d10076b3b908f1045e8dc689f00881d58d6cd0411e240083dfe30d2ef90c8e292868747470733a2f2f66696e6e65792e63616c656e6461722e657465726e69747977616c6c2e636f6df010d565f4b7c8768db9d620719faeb0cff208f1045e8dc688f008d1d67b6d828e79b30083dfe30d2ef90c8e2c2b68747470733a2f2f626f622e6274632e63616c656e6461722e6f70656e74696d657374616d70732e6f7267)

</details>

MarcoFalke commented:
cc @sipsorcery (or anyone else on windows) any idea why this fails appveyor?

ryanofsky commented:
> cc @sipsorcery (or anyone else on windows) any idea why this fails appveyor?

Appveyor links

https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/32000822 (passed)
https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/32021955 (failed)
https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/32023789 (failed)
https://ci.appveyor.com/project/DrahtBot/bitcoin/builds/32039241 (pending)

I think this is the error message

```
src\test_bitcoin.exe -k stdout -e stdout 2> NUL
Command exited with code -1073740791
```

MarcoFalke commented:
Yeah, stuff like this #16976 makes it hard to debug

ryanofsky commented:
The PR was rebased between the passing build and failing build https://github.com/bitcoin/bitcoin/compare/c182fd51..2276339a, but I'd guess probably the new `unregister_all_during_call` unit test is crashing and causing the failure

ryanofsky commented:
Could disable this if appveyor failures cause problems for other PRs:

```diff
--- a/src/test/validationinterface_tests.cpp
+++ b/src/test/validationinterface_tests.cpp
@@ -42,6 +42,9 @@ public:
 // https://github.com/bitcoin/bitcoin/pull/18551
 BOOST_AUTO_TEST_CASE(unregister_all_during_call)
 {
+#ifdef WIN32
+    return;
+#endif
     bool destroyed = false;
 
     CScheduler scheduler;
```

MarcoFalke commented:
appveyor was already failing on every pull, so I think this is not an issue

fanquake commented:
The tests are now failing, and it was caused by this pull. Maybe we can disable as @ryanofsky suggested. I haven't been able to investigate  further:

```
Running 370 test cases...
Assertion failed!

Program: C:\workspace\bitcoin\bin\test_bitcoin.exe
File: validationinterface.cpp, Line 93

Expression: !m_internals
```

ryanofsky commented:
I think I see the problem. The failing assert is in `RegisterBackgroundSignalScheduler` https://github.com/bitcoin/bitcoin/blob/1f70185a809362117a8158e386fdead85728794f/src/validationinterface.cpp#L93

and is probably failing because the test isn't calling `UnregisterBackgroundSignalScheduler` to clean up after itself

ryanofsky commented:
#18563 should probably fix the test failures
